On grad access, check if grad has changed and update stored example grad as needed

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112811

Fixes https://github.com/pytorch/pytorch/issues/112446

This is a doozy of a PR, there's a few important things to keep in mind here:

1) We MUST lift all tensors accessed via attrs to inputs, getattr is a no go in the graph, it violates the aot_autograd contract. Furthermore, aot_autograd does not know how to apply in-place ops to intermediary tensors that are attributes (aka from getattr) anyway. Views from ops are fine.

2) `.grad` access handling in dynamo peeks at the underlying value, the real tensor, because re-piping FakeTensors already made with this fake_mode through builder anew is a no go. 

3) We have no proper mechanism for updating the hint / grapharg.example (the real value in (2) above) midway through trace

Therefore, what we need to do is reconcile the difference in grad stashed on grapharg.example. The easiest way to do this is lazily, upon .grad access, by reading the new value off the right fake tensors. We can then make a tensor using that data as a hint to VariableBuilder to make the right VariableTracker. Note that the example value used here (torch.zeros) in the PR, is a dummy value only used as a tracing hint, it does not leak out into real runtime code. 

Alternatively, we could implement accumulate_grad_ in python...

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

voznesenskym committed:
[hacky] Coerce accumulate_grad_ target grad to not be a None midway through the graph

[ghstack-poisoned]

voznesenskym committed:
Update on "[hacky] Coerce accumulate_grad_ target grad to not be a None midway through the graph"




cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/torch.py:
```diff
@@ -662,6 +662,26 @@ def fn_with_prim_types(x):
+                        var_example = grapharg.example
+                for grapharg in tx.output.graphargs:
+                    if grapharg.source == arg1.source:
+                        grad_example = grapharg.example
```
TODO(voz) - What about inlining? need an example.

voznesenskym replied:
Self note - since this goes to zeros_like, you don't really need a real value. You can do something like this maybe? untested.

```
fgrad = arg1.as_proxy().node.meta['example_value']
var_example.grad = torch.zeros(fgrad.shape, device=fgrad.device)
```

jansel commented on torch/_dynamo/variables/torch.py:
```diff
@@ -662,6 +662,26 @@ def fn_with_prim_types(x):
+                    # In our case, the initial None value is no longer correct.
+                    # By setting it to a dummy zeros tensor, we are guaranteed to have a valid tensor here to coerce the creation of basically
+                    # a dummy TensorVariable instead of a ConstantVariable.
+                    var_example.grad = torch.zeros_like(grad_example)
```
This is wrong, accumulate doesn't allocate a new tensor

voznesenskym replied:
Does this matter? It just becomes a hint to make a tensor variable here. What do you suggest we do instead?

jansel reviewed (changes requested):
This should be done where we do all the other proxy wrapping.  We should just add a check for .grad getting mutated by an op.

voznesenskym commented:
> This should be done where we do all the other proxy wrapping. We should just add a check for .grad getting mutated by an op.

Appreciate the feedback, but this is clearly a draft demonstrating the idea.

voznesenskym committed:
Update on "[hacky] Coerce accumulate_grad_ target grad to not be a None midway through the graph"


Fixes https://github.com/pytorch/pytorch/issues/112446

This is a doozy of a PR, there's a few important things to keep in mind here:

1) We MUST lift all tensors accessed via attrs to inputs, getattr is a no go in the graph, it violates the aot_autograd contract. Furthermore, aot_autograd does not know how to apply in-place ops to intermediary tensors that are attributes (aka from getattr) anyway. Views from ops are fine.

2) `.grad` access handling in dynamo peeks at the underlying value, the real tensor, because re-piping FakeTensors already made with this fake_mode through builder anew is a no go. 

3) We have no proper mechanism for updating the hint / grapharg.example (the real value in (2) above) midway through trace

![image](https://github.com/pytorch/pytorch/assets/4755252/e5bbb8d3-0e37-4279-9c01-0b5bee26759c)


Alternatively, we could implement accumulate_grad_ in python...

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/torch.py:
```diff
@@ -708,6 +717,30 @@ def check_any_unspec(x):
+                                grad_shape_specialized = [
+                                    int(x) for x in new_grad.shape
+                                ]
+                                grapharg.example.grad = torch.zeros(
+                                    grad_shape_specialized, device=new_grad.device
+                                )
```
@jansel what do you suggest instead?

voznesenskym committed:
Update on "[hacky] Coerce accumulate_grad_ target grad to not be a None midway through the graph"


Fixes https://github.com/pytorch/pytorch/issues/112446

This is a doozy of a PR, there's a few important things to keep in mind here:

1) We MUST lift all tensors accessed via attrs to inputs, getattr is a no go in the graph, it violates the aot_autograd contract. Furthermore, aot_autograd does not know how to apply in-place ops to intermediary tensors that are attributes (aka from getattr) anyway. Views from ops are fine.

2) `.grad` access handling in dynamo peeks at the underlying value, the real tensor, because re-piping FakeTensors already made with this fake_mode through builder anew is a no go. 

3) We have no proper mechanism for updating the hint / grapharg.example (the real value in (2) above) midway through trace

![image](https://github.com/pytorch/pytorch/assets/4755252/e5bbb8d3-0e37-4279-9c01-0b5bee26759c)


Alternatively, we could implement accumulate_grad_ in python...

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel reviewed (changes requested):
This seems incorrect and overly complex.  Let chat offline.

voznesenskym commented:
> This seems incorrect and overly complex. Let chat offline.

We can do this lazily. This does it eagerly. The idea is the same regardless. I can PR the lazy version so you can compare.

voznesenskym committed:
Update on "[hacky] Coerce accumulate_grad_ target grad to not be a None midway through the graph"


Fixes https://github.com/pytorch/pytorch/issues/112446

This is a doozy of a PR, there's a few important things to keep in mind here:

1) We MUST lift all tensors accessed via attrs to inputs, getattr is a no go in the graph, it violates the aot_autograd contract. Furthermore, aot_autograd does not know how to apply in-place ops to intermediary tensors that are attributes (aka from getattr) anyway. Views from ops are fine.

2) `.grad` access handling in dynamo peeks at the underlying value, the real tensor, because re-piping FakeTensors already made with this fake_mode through builder anew is a no go. 

3) We have no proper mechanism for updating the hint / grapharg.example (the real value in (2) above) midway through trace

![image](https://github.com/pytorch/pytorch/assets/4755252/e5bbb8d3-0e37-4279-9c01-0b5bee26759c)


Alternatively, we could implement accumulate_grad_ in python...

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "[hacky] Coerce accumulate_grad_ target grad to not be a None midway through the graph"


Fixes https://github.com/pytorch/pytorch/issues/112446

This is a doozy of a PR, there's a few important things to keep in mind here:

1) We MUST lift all tensors accessed via attrs to inputs, getattr is a no go in the graph, it violates the aot_autograd contract. Furthermore, aot_autograd does not know how to apply in-place ops to intermediary tensors that are attributes (aka from getattr) anyway. Views from ops are fine.

2) `.grad` access handling in dynamo peeks at the underlying value, the real tensor, because re-piping FakeTensors already made with this fake_mode through builder anew is a no go. 

3) We have no proper mechanism for updating the hint / grapharg.example (the real value in (2) above) midway through trace

![image](https://github.com/pytorch/pytorch/assets/4755252/e5bbb8d3-0e37-4279-9c01-0b5bee26759c)


Alternatively, we could implement accumulate_grad_ in python...

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/torch.py:
```diff
@@ -749,7 +749,6 @@ def check_any_unspec(x):
                         tx.symbolic_locals[name] = tensor_variable
                 else:
                     unimplemented(f"out variant of {type(kwargs['out'])}")
-
```
self-nit undo

voznesenskym commented:
Test failures seem unrelated.

voznesenskym commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1100,10 +1100,30 @@ def call_getattr(
+                            grad_shape_specialized = [int(x) for x in new_grad.shape]
+                            # We lazily update the grad on the example to its real state as tracked by fake tensor.
+                            # This allocation is fine - it is just a hint. It will not make it to runtime, but it coerces
+                            # the underlying value to always be correct.
+                            grapharg.example.grad = torch.zeros(
+                                grad_shape_specialized, device=new_grad.device
+                            )
```
Ideally, this would use `new_grad` and be done with it, but atm, we cannot, creating a VT via VariableBuilder w/ a fake tensor does not work atm.

voznesenskym commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1100,10 +1100,30 @@ def call_getattr(
+                        new_grad = obj.as_proxy().node.meta["example_value"].grad
+
+                        def _grad_changed(old, new):
+                            if old_grad is None:
```
do we need a new_grad is None path?

voznesenskym replied:
Yes, we do.

voznesenskym committed:
Update on "On grad access, check if grad has changed and update stored example grad as needed "


Fixes https://github.com/pytorch/pytorch/issues/112446

This is a doozy of a PR, there's a few important things to keep in mind here:

1) We MUST lift all tensors accessed via attrs to inputs, getattr is a no go in the graph, it violates the aot_autograd contract. Furthermore, aot_autograd does not know how to apply in-place ops to intermediary tensors that are attributes (aka from getattr) anyway. Views from ops are fine.

2) `.grad` access handling in dynamo peeks at the underlying value, the real tensor, because re-piping FakeTensors already made with this fake_mode through builder anew is a no go. 

3) We have no proper mechanism for updating the hint / grapharg.example (the real value in (2) above) midway through trace

![image](https://github.com/pytorch/pytorch/assets/4755252/e5bbb8d3-0e37-4279-9c01-0b5bee26759c)


Alternatively, we could implement accumulate_grad_ in python...

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented on test/dynamo/test_hooks.py:
```diff
@@ -582,11 +583,13 @@ def fn(x, y):
                     # X goes to x*2 becaue of mul_
                     self.assertEqual(x, torch.tensor([0.5, 0.5, 0.5]) * 2)
+                    # This test proves grad aliasing works -
+                    self.assertEqual(x.grad, b * 5)
```
I cannot easily tell that the equation here is correct. It would be a lot more convincing if you compared the result here to eager mode.

ezyang commented on torch/_dynamo/compiled_autograd.py:
```diff
@@ -147,7 +147,10 @@ def post_acc_grad_hook(self, inputs, hook_id):
         )
         with disable_proxy_modes_tracing():
-            inputs = [maybe_clone(x) for x in inputs]
+            if isinstance(inputs, list):
```
This seems weird. Naively, I would expect an invariant that this is only ever called with a list of inputs. When is it not called with list of inputs? Also cc @jansel

voznesenskym replied:
This is actually never called with a list of inputs. I can remove that.

voznesenskym committed:
Update on "On grad access, check if grad has changed and update stored example grad as needed "


Fixes https://github.com/pytorch/pytorch/issues/112446

This is a doozy of a PR, there's a few important things to keep in mind here:

1) We MUST lift all tensors accessed via attrs to inputs, getattr is a no go in the graph, it violates the aot_autograd contract. Furthermore, aot_autograd does not know how to apply in-place ops to intermediary tensors that are attributes (aka from getattr) anyway. Views from ops are fine.

2) `.grad` access handling in dynamo peeks at the underlying value, the real tensor, because re-piping FakeTensors already made with this fake_mode through builder anew is a no go. 

3) We have no proper mechanism for updating the hint / grapharg.example (the real value in (2) above) midway through trace

Therefore, what we need to do is reconcile the difference in grad stashed on grapharg.example. The easiest way to do this is lazily, upon .grad access, by reading the new value off the right fake tensors. We can then make a tensor using that data as a hint to VariableBuilder to make the right VariableTracker. Note that the example value used here (torch.zeros) in the PR, is a dummy value only used as a tracing hint, it does not leak out into real runtime code. 

Alternatively, we could implement accumulate_grad_ in python...

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented:
Maybe the failures are real? weird. Will rebase and see. Suspecting flaky base.

voznesenskym committed:
Update on "On grad access, check if grad has changed and update stored example grad as needed "


Fixes https://github.com/pytorch/pytorch/issues/112446

This is a doozy of a PR, there's a few important things to keep in mind here:

1) We MUST lift all tensors accessed via attrs to inputs, getattr is a no go in the graph, it violates the aot_autograd contract. Furthermore, aot_autograd does not know how to apply in-place ops to intermediary tensors that are attributes (aka from getattr) anyway. Views from ops are fine.

2) `.grad` access handling in dynamo peeks at the underlying value, the real tensor, because re-piping FakeTensors already made with this fake_mode through builder anew is a no go. 

3) We have no proper mechanism for updating the hint / grapharg.example (the real value in (2) above) midway through trace

Therefore, what we need to do is reconcile the difference in grad stashed on grapharg.example. The easiest way to do this is lazily, upon .grad access, by reading the new value off the right fake tensors. We can then make a tensor using that data as a hint to VariableBuilder to make the right VariableTracker. Note that the example value used here (torch.zeros) in the PR, is a dummy value only used as a tracing hint, it does not leak out into real runtime code. 

Alternatively, we could implement accumulate_grad_ in python...

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "On grad access, check if grad has changed and update stored example grad as needed "


Fixes https://github.com/pytorch/pytorch/issues/112446

This is a doozy of a PR, there's a few important things to keep in mind here:

1) We MUST lift all tensors accessed via attrs to inputs, getattr is a no go in the graph, it violates the aot_autograd contract. Furthermore, aot_autograd does not know how to apply in-place ops to intermediary tensors that are attributes (aka from getattr) anyway. Views from ops are fine.

2) `.grad` access handling in dynamo peeks at the underlying value, the real tensor, because re-piping FakeTensors already made with this fake_mode through builder anew is a no go. 

3) We have no proper mechanism for updating the hint / grapharg.example (the real value in (2) above) midway through trace

Therefore, what we need to do is reconcile the difference in grad stashed on grapharg.example. The easiest way to do this is lazily, upon .grad access, by reading the new value off the right fake tensors. We can then make a tensor using that data as a hint to VariableBuilder to make the right VariableTracker. Note that the example value used here (torch.zeros) in the PR, is a dummy value only used as a tracing hint, it does not leak out into real runtime code. 

Alternatively, we could implement accumulate_grad_ in python...

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "On grad access, check if grad has changed and update stored example grad as needed "


Fixes https://github.com/pytorch/pytorch/issues/112446

This is a doozy of a PR, there's a few important things to keep in mind here:

1) We MUST lift all tensors accessed via attrs to inputs, getattr is a no go in the graph, it violates the aot_autograd contract. Furthermore, aot_autograd does not know how to apply in-place ops to intermediary tensors that are attributes (aka from getattr) anyway. Views from ops are fine.

2) `.grad` access handling in dynamo peeks at the underlying value, the real tensor, because re-piping FakeTensors already made with this fake_mode through builder anew is a no go. 

3) We have no proper mechanism for updating the hint / grapharg.example (the real value in (2) above) midway through trace

Therefore, what we need to do is reconcile the difference in grad stashed on grapharg.example. The easiest way to do this is lazily, upon .grad access, by reading the new value off the right fake tensors. We can then make a tensor using that data as a hint to VariableBuilder to make the right VariableTracker. Note that the example value used here (torch.zeros) in the PR, is a dummy value only used as a tracing hint, it does not leak out into real runtime code. 

Alternatively, we could implement accumulate_grad_ in python...

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented:
@pytorchbot merge

voznesenskym committed:
Update on "On grad access, check if grad has changed and update stored example grad as needed "


Fixes https://github.com/pytorch/pytorch/issues/112446

This is a doozy of a PR, there's a few important things to keep in mind here:

1) We MUST lift all tensors accessed via attrs to inputs, getattr is a no go in the graph, it violates the aot_autograd contract. Furthermore, aot_autograd does not know how to apply in-place ops to intermediary tensors that are attributes (aka from getattr) anyway. Views from ops are fine.

2) `.grad` access handling in dynamo peeks at the underlying value, the real tensor, because re-piping FakeTensors already made with this fake_mode through builder anew is a no go. 

3) We have no proper mechanism for updating the hint / grapharg.example (the real value in (2) above) midway through trace

Therefore, what we need to do is reconcile the difference in grad stashed on grapharg.example. The easiest way to do this is lazily, upon .grad access, by reading the new value off the right fake tensors. We can then make a tensor using that data as a hint to VariableBuilder to make the right VariableTracker. Note that the example value used here (torch.zeros) in the PR, is a dummy value only used as a tracing hint, it does not leak out into real runtime code. 

Alternatively, we could implement accumulate_grad_ in python...

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "On grad access, check if grad has changed and update stored example grad as needed "


Fixes https://github.com/pytorch/pytorch/issues/112446

This is a doozy of a PR, there's a few important things to keep in mind here:

1) We MUST lift all tensors accessed via attrs to inputs, getattr is a no go in the graph, it violates the aot_autograd contract. Furthermore, aot_autograd does not know how to apply in-place ops to intermediary tensors that are attributes (aka from getattr) anyway. Views from ops are fine.

2) `.grad` access handling in dynamo peeks at the underlying value, the real tensor, because re-piping FakeTensors already made with this fake_mode through builder anew is a no go. 

3) We have no proper mechanism for updating the hint / grapharg.example (the real value in (2) above) midway through trace

Therefore, what we need to do is reconcile the difference in grad stashed on grapharg.example. The easiest way to do this is lazily, upon .grad access, by reading the new value off the right fake tensors. We can then make a tensor using that data as a hint to VariableBuilder to make the right VariableTracker. Note that the example value used here (torch.zeros) in the PR, is a dummy value only used as a tracing hint, it does not leak out into real runtime code. 

Alternatively, we could implement accumulate_grad_ in python...

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented:
@pytorchbot merge

JackCaoG commented:
nice let me try this out! I am not sure if this will work given XLA doesn't have lowering for `accumulated_grad_`
